### ISSUE TITLE
Proposed PR :~ unit tests for 'express-api-server'

### ISSUE THREAD

**Issue by nfreear** [axa-group/nlp.js#580]
Hi @jesus-seijas-sp, 

**Is your feature request related to a problem? Please describe.**
Currently, the `express-api-server` package doesn't have any unit tests.

**Describe the solution you'd like**
I'd like to create a pull request that fixes that.

I propose adding the `devDependency` [`supertest`](https://npmjs.com/package/supertest).

To facilitate testing, and following the pattern in [this blog post](https://www.albertgao.xyz/2017/05/24/how-to-test-expressjs-with-jest-and-supertest/), I propose splitting the current `ExpressApiServer` class in two, while maintaining the existing API (both in the one package):

### 1. ExpressApiApp
```js
class ExpressApiApp {
  // ...

  /* Do everything BUT listen to a port!
  */
  initialize() {
    this.app = express();
    this.app.use(cors());
    this.app.use(express.urlencoded({ extended: false }));
    this.app.use(express.json());
    for (let i = 0; i < this.plugins.length; i += 1) {
      this.app.use(this.plugins[i]);
    }
    if (this.settings.serveBot) {
      this.app.use(express.static(path.join(__dirname, './public')));
    }
    for (let i = 0; i < this.routers.length; i += 1) {
      this.app.use(this.settings.apiRoot, this.routers[i]);
    }
    return this.app;
  }
}
```

### 2. ExpressApiServer
```js
class ExpressApiServer ... {
  // ...

  /* Just listen to the port!
  */
  start(input = {}) {
    this.settings.port = input.port || this.settings.port;
    const expressApp = new ExpressApiApp(this.settings, this.plugins, this.routers);
    this.app = expressApp.initialize();

    if (port && port > 0) {
      this.app.listen(port, () => {
        const logger = this.container.get('logger');
        logger.info(`${this.settings.tag} listening on port ${port}!`);
      });
    }
  }
}
```

**Describe alternatives you've considered**
_I've not looked._

**Additional context**
I'm currently up to 6 tests ...

```md
Mxx:nlp.js nXX$ ./node_modules/.bin/jest ./packages/express-api-server --no-coverage

 PASS  packages/express-api-server/test/express-api-app.test.js
  ExpressApiApp
    Constructor
      ✓ Constructor (2 ms)
    Static files should be served
      ✓ The route "/" returns a static HTML page (49 ms)
      ✓ The route "/main.js" returns Javascript (11 ms)
      ✓ The route "/botchat.css" returns a CSS stylesheet (5 ms)
    In isolation, the API gives an error
      ✓ In isolation, the route "/api" returns a "404 Not Found" error (5 ms)
    An API with a router is OK
      ✓ The route "/api/testId/{N}" returns data (7 ms)

Test Suites: 1 passed, 1 total
Tests:       6 passed, 6 total
Snapshots:   0 total
Time:        1.274 s

Ran all test suites matching /.\/packages\/express-api-server/i.
```

Is this proposal OK?

Yours

Nick

**Comment by nfreear** [comment]
Hi @jesus-seijas-sp,

Further notes on my proposal:

 1. Although it would be possible to put the `initialize()` and `start()` functions above in one class, I'm proposing separate `App` and `Server` classes. Let me know if you _do_ want just the one class.

 2. The `supertest` `request().get()` function is asynchronous, so I've been using `async/await` in my tests as its cleaner. Would you prefer me to use `then()` instead, to match the rest of the code-base? (_The fact that you can't use `async/await` when developing NLP.js plugins took me a while to figure out!_)

Any comments/ suggestions welcome!

Thanks,


Nick

**Comment by jesus-seijas-sp** [comment]
Hello! Currently the tests of the PR are failing because the package.json and the package-lock.json are not in sync, it seems that because the supertest@^4.0.2 dependency: 
https://travis-ci.com/github/axa-group/nlp.js/jobs/384355673#L217

**Comment by nfreear** [comment]
Hi @jesus-seijas-sp,

Apologies for the delay ... _ready to merge_ (Pull #586).

N.

**Comment by nfreear** [comment]
Closed by the merge of #586 ... N.
